File: app/src/main/java/com/nirbhor/app/ui/marks/MedicineMark.kt
package com.nirbhor.app.ui.marks

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.clipRect
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * The six medicine marks (README > Medicine Marks). Every medicine is assigned ONE mark at
 * creation; the same mark represents it on every screen (timeline, cabinet, alarm, record, email).
 * It is the primary non-textual identifier for low-literacy users, so it is stored on the medicine
 * record — never derived.
 *
 * Geometry is authored in a 30×30 box and scaled to the requested [size]. Colour is passed in so
 * callers can use the stored markColor, or the lightened variants on the dark alarm background.
 */
enum class MarkShape {
    FilledCircle,   // circle r=13            — default calm
    Ring,           // circle r=11.5, sw 3.5  — default calm
    HalfFilled,     // ring + right half fill — default calm
    RoundedSquare,  // rect 24×24, r=7        — default slate
    Triangle,       // M15 3.5 L27 26 H3 Z    — default ochre
    Capsule;        // rect 22×10, r=5        — default mauve

    companion object {
        fun fromName(name: String?): MarkShape =
            entries.firstOrNull { it.name == name } ?: FilledCircle
    }
}

@Composable
fun MedicineMark(
    shape: MarkShape,
    color: Color,
    modifier: Modifier = Modifier,
    size: Dp = 30.dp,
) {
    Canvas(modifier = modifier.size(size)) {
        val k = this.size.minDimension / 30f // design box is 30×30
        fun p(v: Float) = v * k
        when (shape) {
            MarkShape.FilledCircle ->
                drawCircle(color = color, radius = p(13f), center = Offset(p(15f), p(15f)))

            MarkShape.Ring ->
                drawCircle(
                    color = color, radius = p(11.5f), center = Offset(p(15f), p(15f)),
                    style = Stroke(width = p(3.5f)),
                )

            MarkShape.HalfFilled -> {
                // Right half solid, then the full ring on top.
                clipRect(left = p(15f), top = 0f, right = this.size.width, bottom = this.size.height) {
                    drawCircle(color = color, radius = p(11.5f), center = Offset(p(15f), p(15f)))
                }
                drawCircle(
                    color = color, radius = p(11.5f), center = Offset(p(15f), p(15f)),
                    style = Stroke(width = p(3.5f)),
                )
            }

            MarkShape.RoundedSquare ->
                drawRoundRect(
                    color = color,
                    topLeft = Offset(p(3f), p(3f)),
                    size = Size(p(24f), p(24f)),
                    cornerRadius = CornerRadius(p(7f), p(7f)),
                )

            MarkShape.Triangle ->
                drawPath(
                    color = color,
                    path = Path().apply {
                        moveTo(p(15f), p(3.5f))
                        lineTo(p(27f), p(26f))
                        lineTo(p(3f), p(26f))
                        close()
                    },
                )

            MarkShape.Capsule ->
                drawRoundRect(
                    color = color,
                    topLeft = Offset(p(4f), p(10f)),
                    size = Size(p(22f), p(10f)),
                    cornerRadius = CornerRadius(p(5f), p(5f)),
                )
        }
    }
}

File: lib/ui/marks/medicine_mark.dart
import 'package:flutter/material.dart';

/// The six medicine marks (README > Medicine Marks). Every medicine is assigned ONE mark at
/// creation; the same mark represents it on every screen (timeline, cabinet, alarm, record, email).
/// It is the primary non-textual identifier for low-literacy users, so it is stored on the medicine
/// record — never derived.
///
/// Geometry is authored in a 30×30 box and scaled to the requested size. Colour is passed in so
/// callers can use the stored markColor, or the lightened variants on the dark alarm background.
enum MarkShape {
  filledCircle, // circle r=13            — default calm
  ring, // circle r=11.5, sw 3.5  — default calm
  halfFilled, // ring + right half fill — default calm
  roundedSquare, // rect 24×24, r=7        — default slate
  triangle, // M15 3.5 L27 26 H3 Z    — default ochre
  capsule; // rect 22×10, r=5        — default mauve

  /// Stored form. Kept identical to the Kotlin `MarkShape.name` values so a database copied from
  /// the original app still reads correctly.
  String get storedName => switch (this) {
        MarkShape.filledCircle => 'FilledCircle',
        MarkShape.ring => 'Ring',
        MarkShape.halfFilled => 'HalfFilled',
        MarkShape.roundedSquare => 'RoundedSquare',
        MarkShape.triangle => 'Triangle',
        MarkShape.capsule => 'Capsule',
      };

  static MarkShape fromName(String? name) => MarkShape.values.firstWhere(
        (m) => m.storedName == name,
        orElse: () => MarkShape.filledCircle,
      );
}

class MedicineMark extends StatelessWidget {
  const MedicineMark({
    super.key,
    required this.shape,
    required this.color,
    this.size = 30,
  });

  final MarkShape shape;
  final Color color;
  final double size;

  @override
  Widget build(BuildContext context) => SizedBox(
        width: size,
        height: size,
        child: CustomPaint(painter: _MarkPainter(shape, color)),
      );
}

class _MarkPainter extends CustomPainter {
  _MarkPainter(this.shape, this.color);

  final MarkShape shape;
  final Color color;

  @override
  void paint(Canvas canvas, Size size) {
    final k = (size.width < size.height ? size.width : size.height) / 30; // design box is 30×30
    double p(double v) => v * k;
    final fill = Paint()
      ..color = color
      ..style = PaintingStyle.fill
      ..isAntiAlias = true;
    final stroke = Paint()
      ..color = color
      ..style = PaintingStyle.stroke
      ..strokeWidth = p(3.5)
      ..isAntiAlias = true;
    final center = Offset(p(15), p(15));

    switch (shape) {
      case MarkShape.filledCircle:
        canvas.drawCircle(center, p(13), fill);
      case MarkShape.ring:
        canvas.drawCircle(center, p(11.5), stroke);
      case MarkShape.halfFilled:
        // Right half solid, then the full ring on top.
        canvas.save();
        canvas.clipRect(Rect.fromLTRB(p(15), 0, size.width, size.height));
        canvas.drawCircle(center, p(11.5), fill);
        canvas.restore();
        canvas.drawCircle(center, p(11.5), stroke);
      case MarkShape.roundedSquare:
        canvas.drawRRect(
          RRect.fromRectAndRadius(
            Rect.fromLTWH(p(3), p(3), p(24), p(24)),
            Radius.circular(p(7)),
          ),
          fill,
        );
      case MarkShape.triangle:
        final path = Path()
          ..moveTo(p(15), p(3.5))
          ..lineTo(p(27), p(26))
          ..lineTo(p(3), p(26))
          ..close();
        canvas.drawPath(path, fill);
      case MarkShape.capsule:
        canvas.drawRRect(
          RRect.fromRectAndRadius(
            Rect.fromLTWH(p(4), p(10), p(22), p(10)),
            Radius.circular(p(5)),
          ),
          fill,
        );
    }
  }

  @override
  bool shouldRepaint(_MarkPainter old) => old.shape != shape || old.color != color;
}
